### ISSUE TITLE
Crappy not terminating properly when stopped by a Generator block

### ISSUE THREAD

**Issue by WeisLeDocto** [LaboratoireMecaniqueLille/crappy#12]
Since recent updates in the behavior of the Links, Crappy sometimes doesn't terminate properly when the stop signal is sent by a Generator whose path ended. It then keeps printing `WARNING : Timeout error in pipe send! Link name: <link name>` messages.

It is a minor bug, as it doesn't entail any data loss and it is possible to stop the script by hitting CTRL+C.
This bug does not occur when Crappy is interrupted by the CTRL+C key.

It will likely be fixed when I start working on refactoring and improving the Block, probably in a few months from now.

Weis

**Comment by Elbub** [comment]
Hi !
I've been working with CRAPPy for a few weeks as I'm trying to integrate it to an older Python program that manages a workbench driven through a NI-DAQmx card. I've encountered the same problem, but not only with the Generator block. I've been tinkering a custom Grapher, initially to integrate it to the old software's main window\* and to make crappy stop when you close it or press a stop button. Raising the CrappyStop exception doesn't work, as it passes through matplotlib first and matplotlib doesn't know how to handle it. Thus, I tried to use `crappy.stop()` (or directly `Block.stop_all()`) from different places\*\* and got the same problem : only my EmbeddedGrapher block stopped and all the rest stayed the same.

\* : It turned out to be very hard to achieve, since tk objects aren't picklable, so I set that aside for now.
\*\* : I tried from the `finish()` method and from another custom method called when pressing a button or bound to the matplotlib figure closing event. I tried it from within my script and directly adding the EmbeddedGrpaher subclass to the crappy module.

PS : Je peux aussi vous réexpliquer tout ça en français si besoin.

**Comment by WeisLeDocto** [comment]
Elbub,

As I said in my answer to [your issue](https://github.com/LaboratoireMecaniqueLille/crappy/issues/14#issue-1328572302), I suggest that you use the `1.5.9` version of Crappy or older, available [here](https://github.com/LaboratoireMecaniqueLille/crappy/tree/d207e6cd8418cba56c6216c05e8fcd8cfeda365f) if you want to clone the repo or with `pip install crappy`. The current version of the repo is not stable, while the `1.5.9` or older handles the termination of the blocks in a much smoother way.

It is really hard for me to clearly understand your problem from the explanation you gave. However, I get it that you don't fully understand the logic behind Crappy's blocks termination. So, whenever a class inherits from `crappy.blocks.Block` and calls `super().__init__()` or `crappy.blocks.Block.__init__(self)` in the `__init__` method, it becomes a Crappy block. Crappy then manages the calls to the `prepare`, `finish` and `loop` methods, with the `loop` method called repeatedly until the block termination.

Currently, there are only two ways to stop a Crappy block in a clean way. The first one is to hit the `CTRL+C` key, which will raise a `KeyboardInterrupt` exception in all processes (both Crappy blocks and the other processes). The Crappy blocks will handle it in a nice way, while other processes unrelated to Crappy will have their own way of handling it. Theoretically, it should still stop all processes. Alternatively, any exception raised in a Crappy block will be caught by Crappy, and Crappy will stop all the running blocks. As the exception is caught by Crappy, it may not stop the other processes unrelated to Crappy though, which you would likely have to stop in an other way.

Without more information on your software I cannot help further, you should share code if you need a more precise answer.

Weis

PS : I'm curious to know if the NI-DAQmx InOut that was implemented a long time ago works fine. Feel free to share any problem, suggestion or feedback you would have on the NI-DAQmx implementation !

**Comment by Elbub** [comment]
Oh, it was more of an observation than a problem. Well, not being able to stop other blocks is a problem, but I came here mainly to tell that it's not linked to Generator objects.
I think I understand pretty well how Crappy handles termination of the blocks through the `CrappyStop` exception and `stop_all()` methods, but I'm not really sure about how it handles a SIGINT. 
About the "not terminating properly when Generator stops", it seems that `Block.instances` is emptied when Crappy starts.
I tried to do pretty much the same by appending all the created blocks to a list (since it would avoid garbage collection, unlike a weakset) and defining a function that stops all blocks in that list. That function is called in the CustomGrapher.finish() method. Still, it prints an empty list. That functions is just :
```
def stopping_crappy():
   global list_of_crappy_blocks_used
   print(list_of_crappy_blocks_used)
   for block in list_of_crappy_blocks_used:
      block.stop()
```
I tried with the custom classes in my main script and with them in a separate module. That last attempt was even weirder. I had the list of blocks in the module, and another in the main script than i thought would be just a reference to the very same object (`list_of_crappy_blocks_used= customblocks.list_of_crappy_blocks_used`), thus modifying one would modify both. Then, printing that list from the main before and after `crappy.start()` yielded a complete list of the blocks, but the in-between print (from the function shown above) showed an empty list. Could it be that the list (or `Block.instances` weakset) can't be accessed because they belong to another process (the one managing __main__) ?


Side note : we didn't get any problem with the NI-DAQmx InOut when we tried it. Sadly, we couldn't run much tests, since the card got an unrelated accident that damaged it. We'll get another one soon, so I'll let you know when we'll be able to test it again.
